tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/beam_pipeline.py

```diff
@@ -0,0 +1,339 @@
+
+from tfx.experimental.distributed_inference.graphdef_experiments.subgraph_partitioning import execution_spec
+
+
+@beam.ptransform_fn
+@beam.typehints.with_input_types(Dict[Text, Dict[Text, tf.Tensor]])
+@beam.typehints.with_output_types(Dict[Text, Dict[Text, tf.Tensor]])
+def ExecuteGraph(  # pylint: disable=invalid-name
+    pcoll: beam.pvalue.PCollection,
+    remote_op_name: Text,
+    remote_op_name_to_graph_name: Mapping[Text, Text],
+    graph_name_to_specs: Mapping[Text, List[execution_spec.ExecutionSpec]],
+    graph_to_remote_op_input_name_mapping:
+        Mapping[Text, Mapping[Text, Mapping[Text, Text]]]
+) -> beam.pvalue.PCollection:
+  """Executes a graph.
+
+  Each graph has a list of ExecutionSpecs, in which the order of the list
+  represents the order of execution. An ExecutionSpec can either represent
+  a subgraph layer or a remote op in a remote op layer. When executing a
+  subgraph layer, we can load and execute the subgraph with a beam ParDo.
+  When executing a remote op (which represents another graph), we need to
+  load the remote graph inputs, call ExecuteGraph to switch to that graph,
+  and extract the remote graph output. When executing a remote op, we call the
+  current graph "parent" and the remote graph "child".
+
+  Here, each Beam element is a dictionary from remote op names to a dictionary
+  from tensor names to tensors, or {remote op name: {tensor name: tensor}}.
+  Note that at any time, PColl only stores input tensors and computed tensors.
+  The input PColl should have the input tensor names and values for the graph
+  ready. As we execute the partitioned subgraphs, we add the intermediate
+  output names and values to PColl.
+
+  Args:
+    pcoll: A PCollection of inputs to the graph. Each element is a dictionary
+           from remote op names to a dictionary from tensor names to tensors.
+           Here, element[remote_op_name] contains graph inputs.
+    remote_op_name: The remote op name of the current graph.
+    remote_op_name_to_graph_name:
+      A mapping from remote op names to graph names.
+    graph_name_to_specs:
+      A mapping from graph names to a list of ExecutionSpecs, where the order
+      of the list represents the order of execution.
+
+    # We don't have this information since it was stored in PyFunc's function.
+    graph_to_remote_op_input_name_mapping:
+      A mapping from graph names to remote op names to remote graph placeholder
+      names to parent graph input names.
+      {graph name: {remote op name: {placeholder name: input name}}}.
+
+  Returns:
+    A PCollection of results of this graph. Each element is a dictionary from
+    remote op names to a dictionary from tensor names to tensors. Here,
+    element[remote_op_name] stores graph inputs, intermediate results, and
+    graph outputs.
+  """
+  specs = graph_name_to_specs[remote_op_name_to_graph_name[remote_op_name]]
+
+  for spec in specs:
+    # Construct Beam subgraph for a subgraph layer
```

Comment:
nit: end comments with '.' here and below.

Reply:
Changed.